File: algorithms/c++/15-3sum.cpp
// https://leetcode.com/problems/3sum/
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> sol;
        if (nums.size() < 3) return sol;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        
        for (int i = 0; i < n-2; i++) {
            if (i > 0 && nums[i] == nums[i-1])
                continue;
            int p = i+1, q = n-1;
            while(p < q) {
                int sum = nums[i] + nums[p] + nums[q];
                if (sum ==  0) {
                    sol.push_back({nums[i], nums[p], nums[q]});
                    while(p < n-1 && nums[p] == nums[p+1]) p++;
                    while(q > i && nums[q] == nums[q-1]) q--;
                    p++; q--;
                } else if (sum < 0) {
                    p++;
                } else {
                    q--;
                }
            }
        }
        return sol;
    }
};

int main() {

    return 0;
}

File: algorithms/python/15-3sum.py
# https://leetcode.com/problems/3sum/

from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3: 
            return []
        sol = []
        nums.sort()
        for i in range(0, len(nums)-2):
            if nums[i] > 0:
                return sol
            if i > 0 and nums[i] == nums[i-1]:
                continue
            l = i+1
            r = len(nums)-1
            while(l < r):
                total = nums[i] + nums[l] + nums[r]
                if total == 0:
                    sol.append([nums[i], nums[l], nums[r]])
                    while(l < len(nums)-1 and nums[l+1] == nums[l]):
                        l += 1
                    while(r > i and nums[r-1] == nums[r]):
                        r -= 1
                    l += 1
                    r -= 1
                elif total < 0:
                    l += 1
                else:
                    r -= 1
        return sol
